File: sdk/python/examples/03_persistence.py
#!/usr/bin/env python3
"""KCM Python SDK — Persistence Example.

Demonstrates: save, load, and verify database persistence.
"""

import os
import sys
import tempfile

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "src"))

from kcm import Database, KcmError, ErrorCode


def main() -> None:
    print("=== KCM Python SDK — Persistence Example ===\n")

    path = os.path.join(tempfile.gettempdir(), "kcm_persistence_example.json")

    # --- SAVE DATABASE ---
    print("--- Save Database ---")
    db = Database()
    db.insert(subject=1, predicate=0, object=2, confidence=0.95,
              evidence=1, context=1, version=1, owner=1)
    db.insert(subject=2, predicate=1, object=3, confidence=0.90,
              evidence=2, context=1, version=1, owner=2)
    db.insert(subject=3, predicate=2, object=4, confidence=0.85,
              evidence=3, context=2, version=1, owner=3)
    db.delete(1)  # delete one fact before saving
    print(f"  Facts before save: {db.fact_count()} total, {db.active_fact_count()} active")
    db.save(path)
    print(f"  Saved to {path}")

    # --- VERIFY FILE ---
    print("\n--- Verify Database File ---")
    Database.verify(path)
    print("  Verification passed")

    # --- LOAD INTO NEW DATABASE ---
    print("\n--- Load Into New Database ---")
    db2 = Database()
    db2.load(path)
    print(f"  Loaded: {db2.fact_count()} total, {db2.active_fact_count()} active")
    assert db2.fact_count() == 3
    assert db2.active_fact_count() == 2

    # --- VERIFY DATA INTEGRITY ---
    print("\n--- Verify Data Integrity ---")
    all_facts = db2.query_all()
    for f in all_facts:
        print(f"  row_id={f['row_id']}  s={f['subject']}  p={f['predicate']}  o={f['object']}")
    assert len(all_facts) == 2

    # --- SAVE-LOAD ROUND TRIP ---
    print("\n--- Save-Load Round Trip ---")
    db2.insert(subject=10, predicate=0, object=20, confidence=0.99)
    db2.save(path)
    db3 = Database()
    db3.load(path)
    print(f"  Round-trip: {db3.fact_count()} total, {db3.active_fact_count()} active")
    assert db3.fact_count() == 4
    assert db3.active_fact_count() == 3

    # --- CLEANUP ---
    os.unlink(path)
    db.close()
    db2.close()
    db3.close()
    print("\n=== All persistence operations completed ===")


if __name__ == "__main__":
    main()

File: sdk/javascript/examples/03_persistence.js
#!/usr/bin/env node
"use strict";

/**
 * KCM JavaScript SDK — Persistence Example.
 *
 * Demonstrates: save, load, and verify database persistence.
 * Note: The JS reference implementation throws IO errors for save/load/verify.
 * This example shows the expected error handling pattern.
 */

const { Database, KcmError, ErrorCode } = require('../src/index.js');

console.log("=== KCM JavaScript SDK — Persistence Example ===\n");

const db = new Database();

// Insert facts
db.insert({
    subject: 1, predicate: 0, object: 2, confidence: 0.95,
    evidence: 1, timestamp: 0, context: 1, version: 1, priority: 0, owner: 1,
});
db.insert({
    subject: 2, predicate: 1, object: 3, confidence: 0.90,
    evidence: 2, timestamp: 0, context: 1, version: 1, priority: 0, owner: 2,
});
db.insert({
    subject: 3, predicate: 2, object: 4, confidence: 0.85,
    evidence: 3, timestamp: 0, context: 2, version: 1, priority: 0, owner: 3,
});
console.log(`Facts: ${db.factCount()} total, ${db.activeFactCount()} active\n`);

// --- SAVE (reference implementation) ---
console.log("--- Save Database ---");
try {
    db.save("/tmp/kcm_example.db");
    console.log("  Database saved");
} catch (e) {
    if (e instanceof KcmError) {
        console.log(`  save() threw KcmError: code=${ErrorCode[e.code]}, message="${e.message}"`);
        console.log("  (This is expected for the reference implementation)");
    }
}

// --- LOAD (reference implementation) ---
console.log("\n--- Load Database ---");
try {
    const db2 = new Database();
    db2.load("/tmp/kcm_example.db");
    console.log(`  Loaded: ${db2.factCount()} facts`);
    db2.close();
} catch (e) {
    if (e instanceof KcmError) {
        console.log(`  load() threw KcmError: code=${ErrorCode[e.code]}, message="${e.message}"`);
        console.log("  (This is expected for the reference implementation)");
    }
}

// --- VERIFY (reference implementation) ---
console.log("\n--- Verify Database ---");
try {
    Database.verify("/tmp/kcm_example.db");
    console.log("  Database verified");
} catch (e) {
    if (e instanceof KcmError) {
        console.log(`  verify() threw KcmError: code=${ErrorCode[e.code]}, message="${e.message}"`);
        console.log("  (This is expected for the reference implementation)");
    }
}

// --- VERIFY NON-EXISTENT FILE ---
console.log("\n--- Verify Non-Existent File ---");
try {
    Database.verify("/nonexistent/path/db.kcm");
    console.log("  FAIL: Should have thrown");
} catch (e) {
    if (e instanceof KcmError) {
        console.log(`  Caught: code=${ErrorCode[e.code]}, message="${e.message}"`);
    }
}

// --- LOAD NON-EXISTENT FILE ---
console.log("\n--- Load Non-Existent File ---");
try {
    db.load("/nonexistent/path/db.kcm");
    console.log("  FAIL: Should have thrown");
} catch (e) {
    if (e instanceof KcmError) {
        console.log(`  Caught: code=${ErrorCode[e.code]}, message="${e.message}"`);
    }
}

db.close();
console.log("\n=== All persistence operations completed ===");
